### ISSUE TITLE
exception forwarding in from_thread.run

### ISSUE THREAD

**Issue by PepperoniPingu** [python-trio/trio#3039]
Hi!

Use trio a ton, great library! Thanks for your hard work!

I have an application where I monitor usb insertions through pyudev. The pyudev [MonitorObserver](https://pyudev.readthedocs.io/en/latest/api/pyudev.html#pyudev.MonitorObserver) registers callbacks to udev events that are run in a separate thread. In the callback I used `trio.from_thread.run(func)` to make it work with trio. The problem with this is that if `func` raises an exception, the exception isn't raised in to the main thread but rather gets sent back to the `MonitorObserver` thread which crashes silently without notifying the main thread. 

To solve this I used a memory channel to send the exceptions to a listener task in the trio thread. But this was kind of an awkward solution and it would be great if `from_thread.run` had some kind of flag to make the exception raise in the trio run loop.

What are your thoughts?

regards

**Comment by A5rocks** [comment]
You can probably do it simpler by `trio.from_thread.run(outcome.run, func).unwrap()`, but I don't have much of an opinion on the broader feature request.

**Comment by A5rocks** [comment]
Sorry, I totally misunderstood! If I understand you right, you have two threads:
 - the trio thread
 - a `MonitorObserver` thread you created and passed in the trio token

Then in your `MonitorObserver` callback (which runs in the `MonitorObserver` thread) you run `trio.from_thread.run` which can error. I think the main issue is that... this just isn't structured. There's no task to raise this exception from.

To raise an exception in the trio thread you can use `trio.lowlevel.spawn_system_task`, of note:
> If a system task raises an exception, then it’s converted into a [TrioInternalError](https://trio.readthedocs.io/en/stable/reference-core.html#trio.TrioInternalError) and all tasks are cancelled. If you write a system task, you should be careful to make sure it doesn’t crash.

---

Basically I don't think this is something trio can do while being structured concurrency. There is no relevant task to inject the exception in, and unlike `KeyboardInterrupt` it doesn't make sense to inject into the currently running task. I'm going to close this, let me know if I misunderstood yet again or if you think I'm wrong.

**Comment by TeamSpen210** [comment]
What you could do instead is to just have the observer thread immediately send the `Device` values via a memory channel, then do all your processing in a task. The memory channel itself isn't going to raise exceptions, so the observer won't crash. But if the task fails that'll produce exceptions in the right places.
